service/controller/v22/resource/s3bucket/delete.go

```diff
@@ -23,62 +35,92 @@ func (r *Resource) ApplyDeleteChange(ctx context.Context, obj, deleteChange inte
 		return microerror.Mask(err)
 	}
 
-	for _, bucketInput := range bucketsInput {
-		if bucketInput.Name == "" {
-			continue
-		}
+	g, ctx := errgroup.WithContext(ctx)
 
-		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("deleting S3 bucket %q", bucketInput.Name))
+	for _, b := range bucketsInput {
+		bucketName := b.Name
 
-		var repeat bool
-		for {
-			i := &s3.ListObjectsV2Input{
-				Bucket: aws.String(bucketInput.Name),
-			}
-			o, err := sc.AWSClient.S3.ListObjectsV2(i)
-			if err != nil {
-				return microerror.Mask(err)
-			}
-			if o.IsTruncated != nil && *o.IsTruncated {
-				repeat = true
-			}
-			if len(o.Contents) == 0 {
-				break
+		g.Go(func() error {
+			if bucketName == "" {
+				return nil
 			}
 
-			r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("deleting %d objects", len(o.Contents)))
+			r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("deleting S3 bucket %q", bucketName))
 
-			for _, o := range o.Contents {
-				i := &s3.DeleteObjectInput{
-					Bucket: aws.String(bucketInput.Name),
-					Key:    o.Key,
+			var count int
+			var repeat bool
+			for {
```

Comment:
Why not `for repeat {` and then remove the condition from the end? When `ListObjectsV2()` doesn't return objects this breaks already and when loop count is up it sets repeat to false. I always get a heart attack when I see normal business logic and there's infinite loop. :slightly_smiling_face:

Reply:
Good call. Reason here is not thinking outside the box and the code growing over time. Will check on it.

Reply:
I checked and removed `repeat` completely. The loop is still "infinite" but the important thing here for me was to reduce the amount of control flow primitives. We do not use booleans and `break` statements, but only `break` statements.